pkg/tfshim/sdk-v2/upgrade_state.go

```diff
@@ -32,61 +34,38 @@ func upgradeResourceState(ctx context.Context, p *schema.Provider, res *schema.R
 		if err != nil {
 			return nil, err
 		}
-		version, hasVersion = int(v), true
+		version, hasVersion = v, true
 	}
 
-	// First, build a JSON state from the InstanceState.
-	json, version, err := schema.UpgradeFlatmapState(ctx, version, m, res, p.Meta())
-	if err != nil {
-		return nil, err
+	var (
+		rawState cty.Value
+		err      error
+	)
+	switch GetInstanceStateStrategy(v2Resource{res}) {
+	// Opt-out of normalization under a flag. Normalization replaces nulls with empty blocks for the state value,
+	// but when processing config the codebase calls NewResourceConfigShimmed which in turn calls
+	// ConfigValueFromHCL2Block which removes empty blocks. This leads to non-empty diffs.
+	case CtyInstanceState:
```

Comment:
We can start with a PR that removes CtyInstanceState flag from the codebase. org:pulumi search shows that it's' no longer deployed anywhere. I think the use case in AWS was superceded by WithPlanResourceChange.

Reply:
Done. That makes this PR simpler.